test/xbgpu/test_engine.py

```diff
@@ -1452,3 +1573,150 @@ def get_stream_status(stream_name: str) -> bool:
+        check the corresponding streams only have partial data transmission.
+        Also ensure that data is completely received for data streams that did
+        not receive a `?capture-stop` request.
+
+        This test is carried out in combinations to fully exercise stop logic in
+        both X- and BPipelines.
+        """
+        assert n_x_streams_to_stop <= len(corrprod_outputs)
+        assert n_b_streams_to_stop <= len(beam_outputs)
+        n_batches = heap_accumulation_threshold[0]
+        # NOTE This value is arbitrarily chosen, but must be less than
+        # `n_batches` / `HEAPS_PER_FENGINE_PER_CHUNK` to actually take effect.
+        # A stop index <= 0 (zero) is treated as invalid/an indication to proceed
+        # as normal for the following reasons:
+        # - Using index zero (0) to obtain a subset of results yields an empty array.
+        # - Using a negative index, while valid, results in a non-deterministic
+        #   point of issue for the capture-stop request(s). i.e. it depends on
+        #   `n_batches`, which is not ideal.
+        # NOTE: Like the XBEngine receiver, the BPipeline send-stream transmits
+        # chunks that contain `HEAPS_PER_FENGINE_PER_CHUNK`. Consequently, the
+        # `?capture-stop` request must be issued at least one send-chunk before
+        # data transmission is over to ensure some strength in the test.
+        capture_stop_chunk_index = 10
+        assert capture_stop_chunk_index < n_batches // HEAPS_PER_FENGINE_PER_CHUNK
+        corrprod_capture_stop_heap_indices = [0] * len(corrprod_outputs)
+        if n_x_streams_to_stop > 0:
+            stopped_corrprods = [corrprod_output for corrprod_output in corrprod_outputs[:n_x_streams_to_stop]]
+            capture_stop_corrprods = []
+            for stopped_corrprod in stopped_corrprods:
+                stream_index = corrprod_outputs.index(stopped_corrprod)
+                # NOTE: Data is received at the heap level. As a result, we multiply the chunk
+                # ID at which `?capture-stop` was issued by `HEAPS_PER_FENGINE_PER_CHUNK` to
+                # get the specific heap ID at which data ceased transmitting for a stopped
+                # stream.
+                corrprod_capture_stop_heap_indices[stream_index] = (
+                    capture_stop_chunk_index * HEAPS_PER_FENGINE_PER_CHUNK
+                )
+                capture_stop_corrprods.append(("capture-stop", stopped_corrprod.name))
+            self._patch_method(
+                monkeypatch,
+                "_get_in_item",
+                XPipeline,
+                capture_stop_chunk_index,
+                client,
+                capture_stop_corrprods,
+            )
+
+        beam_capture_stop_heap_indices = [0] * len(beam_outputs)
+        if n_b_streams_to_stop > 0:
+            stopped_beams = [beam_output for beam_output in beam_outputs[:n_b_streams_to_stop]]
```

Comment:
This is a case where randomness might not be a bad idea (take a random selection of outputs instead of a prefix).

Reply:
Still to be addressed.